packages/doenetml-worker-rust/lib-doenetml-core/src/dast/ref_resolve/build_resolver.rs

```diff
@@ -276,24 +288,79 @@ impl Resolver {
 
                     // Add `element` to the name map of `parent`, creating an ambiguous reference
                     // if its `name_with_source` is already in the name map
-                    descendant_names[parent_idx_plus_1]
-                        .get_mut(&name_with_source)
-                        .map(|x| {
-                            *x = match x {
+                    match descendant_names[parent_idx_plus_1].get_mut(&name_with_source) {
+                        Some(x) => {
+                            match x {
                                 // There is already something sharing the `name_with_source` with the current element
-                                // so we mark it as ambiguous
-                                Ref::Unique(idx) => Ref::Ambiguous(vec![*idx, element.idx]),
+                                Ref::Unique(idx) => {
+                                    if *idx == previous_parent {
+                                        // if `name_with_source` resolves to the `previous_parent`,
+                                        // then the parent supersedes the name of the descendant.
+                                        // We don't change the name map, and we stop looping to additional parents
+                                        break;
+                                    } else if let Some(Ref::Unique(descendent_idx)) =
+                                        descendant_ref_of_own_name
```

Comment:
I think we can `.get` `descendant_ref_of_own_name` right here and avoid looking stuff up in some cases. Maybe we can also avoid the `cloned()`

Reply:
We can't because we have a open mutable reference to `descendant_names[parent_idx_plus_1]` starting at line `291`. That's why I had to move it outside and add a `cloned()`. Almost all the time `descendant_ref_of_own_name` is going to be `None`, so I hope it doesn't slow things down too much. And, if it isn't `None`, then the algorithm will reach one of the two places where it is used. I can't think of a better way to do it given rust borrowing rules.